PR to implement issue #2.

hashhar opened the pull request:
This commit adds a context menu entry to the explorer and an associated
command to open the file (which was clicked on) in an external program.

The external program can be configured using the extension's
configuration under the key `externalEditor`. The key can accept full
paths to the tools or just the name of the program if it is present in
PATH. Arguments to be passed to the program can also be specified in the
accompanying configuration under the key `externalEditorArgs`.

Example configuration:

```
"vscode-opn.externalEditor": "emacs",
"vscode-opn.externalEditorArgs": [
    "+10",
    "-nbc"
]
```

hashhar commented:
I think the linter is having some issues.

I have tested this code very rigorously to make sure that it never fails ungracefully.

I took care of:
1. If the click did not happen on a file.
2. If the config file has issues with the two new keys.

EDIT: Fixed the linter issues.

hashhar commented:
@lfurzewaddock Just wanted to remind you that I have submitted a PR.

lfurzewaddock commented:
Apologies @hashhar, I've been away on holiday. Please give me a few days to catch up.

lfurzewaddock commented:
Before I go through any code, I'd like to check what the intended behaviour is. When using your 'Open file...' context menu, is it only possible to set a single program/editor to handle all languages, (therefore file types)?

If so, this doesn't fit well with the existing functionality. Would it be possible to use the existing optional settings per language [JSON (vscode-opn.perLang)](https://github.com/lfurzewaddock/vscode-opn#customise-behaviour-per-language-optional), which set both the program/editor and additional arguments to send to the program/editor per language? Also, as these settings are optional, if there are no settings for the relevant language set, the OS default program/editor should open.

hashhar commented:
Currently I have added a configuration setting called `externalEditor` which is used for all programs.
I think what you're suggesting is better but it can be improved further if we use the language specific settings if they are present and use the `externalEditor` otherwise. If `externalEditor` is also blank, we use the OS defaults.

If you like this approach I can implement it and send you a PR. But it may take some time as my campus placements are on right now and will be until this Saturday.

lfurzewaddock commented:
I can see why you might want a 'super' program/editor to handle various languages, without explicitly setting the same editor for several languages, so how about adding 'defaultAll' as a pseudo language option to [JSON (vscode-opn.perLang)](https://github.com/lfurzewaddock/vscode-opn#customise-behaviour-per-language-optional) which could act as a fall through 'catch all' option?

That way related settings are in the same place and the feature is easier to communicate to users.

I'm OK to wait, I can only fit this around other things too.

hashhar commented:
Could you explain what you meant by defaultAll with an example JSON config?

EDIT: Ok, I understand now. I'll start working later.

lfurzewaddock commented:
See below the original example in the docs with a 3rd option added to the `opnOptions` array. Notice the extra optional options;
- `isUseWebServer` to open file on a local webserver 
- `isUseFsPath` to force use of file system path over URI path
- `isWaitForAppExit` is a specific argument passed to the Node module dependency [opn](https://github.com/sindresorhus/opn)

```

"vscode-opn.perLang": {
    "opnOptions": [
      {
        "forLang": "html",
        "openInApp": "chrome",
        "openInAppArgs": ["--incognito"],
        "isUseWebServer": true,
        "isUseFsPath": false,
        "isWaitForAppExit": true
      },
      {
        "forLang": "plaintext",
        "openInApp": "notepad++",
        "openInAppArgs": [],
        "isUseWebServer": false,
        "isUseFsPath": true,
        "isWaitForAppExit": true
      },
      {
        "forLang": "defaultAll",
        "openInApp": "emacs",
        "openInAppArgs": ["+10", "-nbc"],
        "isUseWebServer": false,
        "isUseFsPath": false,
        "isWaitForAppExit": true
      }
    ]
  },
  "vscode-opn.webServerProtocol": "http",
  "vscode-opn.webServerHostname": "localhost",
  "vscode-opn.webServerPort": 8080

```

hashhar committed:
Fixes #2. Set a default editor.

This commit adds a new pseudo-language id which can be used to set a
default editor to be used in the absence of a configuration for the
current language id.

It also adds a context menu entry to launch the command.

In the absence of any language configuration, the default editor of the
OS will be used.

Example configuration:
```
"vscode-opn.perLang": {
    "opnOptions": [
	{
	    "forLang": "html",
	    "openInApp": "chrome",
	},
	{
	    "forLang": "default",
	    "openInApp": "notepad",
	    "isUseFsPath": true
	}
    ]
}
```

hashhar commented:
@lfurzewaddock I finally got the time to do this. I added a pseudo-language id called `default` which can be used to set options for unconfigured languages. I still have to test it I guess.

I also got around to updating the dependencies and targeting vscode 1.0.

Enjoy a very merry Christmas.

hashhar commented:
@lfurzewaddock Just checking in to remind you about this.

lfurzewaddock commented:
Hi @hashhar, sorry for the delay.

I ran your version locally in debug mode, leaving the default settings and an error is raised when I right click a file and choose option "Open in External Editor".

> Windows cannot find 'null'. Make sure you've typed the name correctly, then try again.

 I'm guessing this is related to the  "vscode-opn.externalEditor" default setting.

lfurzewaddock reviewed (commented):
Is this code duplication required? Why is it not possible to modify the existing function 'disposable' so you can call that from the menu explorer/context command?

lfurzewaddock commented:
This is JavaScript Repo, currently using ES5. Please do not include any 'TypeScript' dependencies/cruft.

@hashhar, I think I should implement your solution as it's a good/useful feature. If I do I will give you credit.

hashhar commented:
@lfurzewaddock Ok, I see what happenned. I did not implement this fully. And was new to VSCode at the time when I started working on this. I can send you a clean ES5 update to this PR sometime by tomorrow night. Or you can implement it yourself if you want.

lfurzewaddock commented:
@hashhar, if you want to submit another PR, I'll review it, but I'd like to see a minimal set of changes so it is easy for me to review and accept.

hashhar commented:
@lfurzewaddock Done. This adds a new pseudo language identifier like we discussed above. **Unfortunately I couldn't figure out how to add a default value for it or add a schema to validate it.**

This also adds a context menu entry to both the editor and the file explorer for the command.

If no language specific configuration is provided then we fall back to the OS defaults.

hashhar commented:
@lfurzewaddock Should I update the dependencies too to get rid of the BitHound errors.

lfurzewaddock commented:
@hashhar, no need. I'm going to make quite a few changes soon.

lfurzewaddock reviewed (changes requested):
Unexpected Behaviour

If I right click on a file a context menu opens for that file. From this context menu when 'Open' is clicked, I would expect the file which had been selected by the the right click to open rather the current active file in the editor.

Please remove your changes from the file 'extension.js' and 'jsconfig.json' and resubmit your pull request so I can merge it.

hashhar commented:
I'll get working on this. Sorry I saw this so late.

hashhar committed:
Fixes #2 and regression in previous commit.

This commit fixes the wrong behaviour of opening the active file instead
of the file from which the context-menu action was invoked.

Now this only passes a `TextDocument` instead of a complete `TextEditor`
to other functions (mainly in `fileService.js`) to help reduce
complexity and allow code-reuse.

This also rolled back the change in `jsconfig.json`.
